### PR TITLE
Fixes & Improvements (previously support components) 

### LINKED ISSUE #24: Use discord-api-types as a dependency
### Summary

We should not maintain the Discord objects that represent the Discord API. Instead, we should use [discord-api-types](https://discord-api-types.dev/).

### Intended Outcome

The outcome should be:

- Adapt our core files to use the `discord-api-types` types, interfaces, and whatnot.
- Remove the `types.ts` file.

### How will it work?

We should add in the project's dependency tree the latest version of `discord-api-types`.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### LINKED ISSUE #25: Support Message Components
### Summary

Discord release not long ago the possibility to add components to the messages. For more information, take a look at the [Discord Developer Portal - Message Components](https://discord.com/developers/docs/interactions/message-components).

This framework should certainly allow easy development of message components as it does for application commands.

### Intended Outcome

The message components should be handled the same way as the application commands do.

### How will it work?

The `Application` type and the `createApplicationCommandHandler` should support components in them.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by Inrixia**
Fixes #24
Fixes #25 

Opening this PR to begin discussion on this feature to add support for components. This is mostly done though re https://github.com/alvgaona/cloudflare-discord-bot/pull/13 and https://github.com/alvgaona/cloudflare-discord-bot/pull/12

I'll need to update formatting first. I intend to setup a eslinf config that closely mirrors the original files format and use that since I have vs setup to properly format/fix files on saving.

### CODE REVIEW

**Comment by Inrixia** [comment]
I also need to move types into the proper file and I want to check if there is a better source for the discord api types (like discord.js) as they are very incomplete

**Comment by Inrixia** [comment]
> > I also need to move types into the proper file and I want to check if there is a better source for the discord api types (like discord.js) as they are very incomplete
> 
> @Inrixia all the types should be in here: https://discord.com/developers/docs

Yep see https://github.com/alvgaona/cloudflare-discord-bot/issues/16 I've already made good headway on switching to the official api types package, it's all uncommitted on this branch

**Comment by alvgaona** [comment]
> > > I also need to move types into the proper file and I want to check if there is a better source for the discord api types (like discord.js) as they are very incomplete
> > 
> > 
> > @Inrixia all the types should be in here: https://discord.com/developers/docs
> 
> Yep see #16 I've already made good headway on switching to the official api types package, it's all uncommitted on this branch

Excellent. I've assigned that issue to you. Feel free to work it out.

**Comment by Inrixia** [comment]
This has now been updated to be upstream from the latest merges to main on this repo.

This PR does two main things, adds a components field for people to add handlers for and changes behind the scenes handling for commands to use a dict for easy lookup instead of an array (which had to be searched every request).

**Comment by alvgaona** [comment]
Hi, @Inrixia my apologies for delaying this. I promise this week I'll spend some time on it. I'll get back to you.

**Comment by Inrixia** [comment]
@alvgaona This branch has tons of fixes including switching to new types mentioned in https://github.com/alvgaona/cloudflare-discord-bot/issues/16 and support for components. I've been updating it as I encounter issues while using it for other projects

File: src/permissions.ts [hunk 1]
```diff
@@ -1,11 +1,13 @@
...
 export class Permissions {
-  constructor(private types: PermissionType[]) {}
+  constructor(private types: PermissionFlags[]) {}
 
   compute() {
-    let permission = 0;
-    this.types.forEach((type) => (permission += type));
+    let permission = 0n;
+    this.types.forEach((type) => (permission += PermissionFlagsBits[type]));
+    console.log(permission, String(permission));
```
Comment (alvgaona) [inline src/permissions.ts:10]: @Inrixia do we want to log this?

**Review by alvgaona — APPROVED** [review]
This is looking pretty neat!

File: src/permissions.ts (hunk 1 above)
Comment (Inrixia) [reply src/permissions.ts:10]: Oops, that's leftover from me testing. I'll remove it

**Comment by alvgaona** [comment]
@Inrixia feel free to merge it if you think this PR's done.

**Comment by Inrixia** [comment]
This pr has some breaking changes afik, so maybe just double check @alvgaona

**Comment by alvgaona** [comment]
> This pr has some breaking changes afik, so maybe just double check @alvgaona

Let me run a smoke test then.

File: src/handler.ts [hunk 2]
```diff
@@ -1,32 +1,44 @@
...
+import { CommandInteractionHandlerWithData, interaction, InteractionDataType } from "./interaction";
 import { Permissions } from "./permissions";
-import { ApplicationCommand, InteractionHandler } from "./types";
+
+import type { CommandInteractionHandler, ComponentInteractionHandler } from "./interaction";
+import type { ApplicationCommand } from ".";
 
 const router = Router();
 
+export type Command<DataType extends InteractionDataType | void = void> = [
```
Comment (alvgaona) [inline src/handler.ts:12]: @Inrixia how do we create a command out of this type? Do you have an example?

File: src/handler.ts [hunk 3]
```diff
@@ -1,32 +1,44 @@
...
+];
+
 export type Application = {
   applicationId: string;
   applicationSecret: string;
   publicKey: string;
   guildId?: string;
-  commands: [ApplicationCommand, InteractionHandler][];
+  commands: Command<any>[];
+  components?: Record<string, ComponentInteractionHandler>;
```
Comment (alvgaona) [inline src/handler.ts:23]: Any example of how to use the components in this case?

File: src/interaction.ts [hunk 4]
```diff
@@ -1,61 +1,88 @@
...
+}
 
-    if (!isValid) throw new Error("Invalid request");
-  };
+type InteractionDataLookup = {
+  [InteractionDataType.ChatInput]: CommandInteractionWithData<ChatInputData>;
+  [InteractionDataType.ContextMenu]: CommandInteractionWithData<MenuInteractionData>;
+};
+
+export type InteractionResponse = Promise<APIInteractionResponse> | APIInteractionResponse;
```
Comment (alvgaona) [inline src/interaction.ts:28]: How is this type used in terms of a client of this library?

**Review by alvgaona — CHANGES_REQUESTED** [review]
Hi @Inrixia, sorry it took me a while to look at this in depth. 

I feel like I need an example from a client's point of view. This PR touches the core of the library and we need to look at this from that point of view as well in this PR. I've tried to set up my own bot off of this PR but had some doubts about it, seems like there are missing types that we could export in order to properly set up commands for instance when writing a bot.

Do you think you could set up a "hello world" bot? It'd be very useful as we could harness that piece of code to set up testing.

**Comment by Inrixia** [comment]
Hey no problem on the delay. Will give this a proper look over when I get time but for now you can see an example bot using this here: https://github.com/Inrixia/LostMokoko

I'll join the discord server so feel free to message me questions there or here.

File: src/handler.ts (hunk 2 above)
Comment (Inrixia) [reply src/handler.ts:12]: https://github.com/Inrixia/LostMokoko/blob/deb098244c4b5868632b1ad2f879a9941933a957/src/reactMentions.ts#L21

File: src/handler.ts (hunk 3 above)
Comment (Inrixia) [reply src/handler.ts:23]: I had a test command which returned a message with a button a user could click in this commit https://github.com/Inrixia/LostMokoko/blob/5065d55612ec136ef7360646da17be965259cd6f/src/hello.ts

Note how the `custom_id` is the same in the command definition
https://github.com/Inrixia/LostMokoko/blob/5065d55612ec136ef7360646da17be965259cd6f/src/hello.ts#L38-L44

and the component definition
https://github.com/Inrixia/LostMokoko/blob/5065d55612ec136ef7360646da17be965259cd6f/src/hello.ts#L51-L62

File: src/interaction.ts (hunk 4 above)
Comment (Inrixia) [reply src/interaction.ts:28]: Assuming you are referring to `InteractionResponse` that type is the expected type any interaction handler will return.

For example this returned object conforms to the `InteractionResponse` type https://github.com/Inrixia/LostMokoko/blob/deb098244c4b5868632b1ad2f879a9941933a957/src/reactMentions.ts#L101-L106

Its specifically this response type
https://github.com/discordjs/discord-api-types/blob/35b17dcbf894b8edf0edbc4c831932192266461e/payloads/v10/_interactions/responses.ts#L48-L51

**Comment by Inrixia** [comment]
@alvgaona Have a look at my responses and the example repo. Can discuss further if you like

File: package.json [hunk 5]
```diff
@@ -29,6 +29,7 @@
 	},
 	"dependencies": {
 		"buffer": "^6.0.3",
+		"discord-api-types": "^0.33.5",
```
Comment (alvgaona) [inline package.json:32]: Let's bump this version up a little bit, to the latest, if possible.

File: src/interaction.ts [hunk 6]
```diff
@@ -1,61 +1,88 @@
...
+import type { DictCommands } from "./handler";
+import {
+  APIMessageComponentInteraction as ComponentInteraction,
+  APIApplicationCommandInteraction as CommandInteraction,
+  APIApplicationCommandInteractionWrapper as CommandInteractionWithData,
+  APIInteractionResponse,
+  APIInteraction,
+  InteractionType,
+  APIChatInputApplicationCommandInteractionData as ChatInputData,
+  APIContextMenuInteractionData as MenuInteractionData,
```
Comment (alvgaona) [inline src/interaction.ts:13]: Can you remove the `as`? I know the names are a bit longer but it's much clear.

File: src/interaction.ts [hunk 7]
```diff
@@ -1,61 +1,88 @@
...
+} from "discord-api-types/payloads";
 
-    const isValid = nacl.sign.detached.verify(Buffer.from(timestamp + JSON.stringify(body)), Buffer.from(signature, "hex"), Buffer.from(publicKey, "hex"));
+export enum InteractionDataType {
+  // APIChatInputApplicationCommandInteractionData
+  ChatInput,
+  // APIContextMenuInteractionData
+  ContextMenu,
```
Comment (alvgaona) [inline src/interaction.ts:20]: Same here. Let's use `discord-api-types` names directly.

File: src/setup.ts [hunk 8]
```diff
@@ -1,93 +1,53 @@
...
-    return `${url}/guilds/${guildId}/commands`;
-  }
-
-  return `${url}/commands`;
+  if (guildId !== undefined) return RouteBases.api + Routes.applicationGuildCommands(applicationId, guildId);
+  return RouteBases.api + Routes.applicationCommands(applicationId);
 };
 
-const deleteExistingCommands = async (applicationId: string, bearer: any, guildId?: string): Promise<void> => {
+const deleteExistingCommands = async (applicationId: string, applicationSecret: string, guildId?: string): Promise<void> => {
```
Comment (alvgaona) [inline src/setup.ts:11]:
```suggestion
const deleteExistingCommands = async (applicationId: string, botToken: string, guildId?: string): Promise<void> => {
```

As it's not anymore the `CLIENT_SECRET` used in here, we're using the Bot's token.

File: src/setup.ts [hunk 9]
```diff
@@ -1,93 +1,53 @@
...
 
-  const request = new Request(url, {
+  const commands = await fetch(url, {
     method: "GET",
-    headers: { "Content-Type": "application/json", Authorization: `Bearer ${bearer}` },
-  });
-
-  const response = await fetch(request);
-  const commands = await response.json();
+    headers: { "Content-Type": "application/json", Authorization: `Bot ${applicationSecret}` },
```
Comment (alvgaona) [inline src/setup.ts:16]:
```suggestion
    headers: { "Content-Type": "application/json", Authorization: `Bot ${token}` },
```

File: src/setup.ts [hunk 10]
```diff
@@ -1,93 +1,53 @@
...
   await Promise.all(
-    commands.map((command: ApplicationCommand & { id: string; application_id: string }) => {
-      return fetch(`${url}/${command.id}`, {
+    commands.map((command) =>
+      fetch(`${url}/${command.id}`, {
         method: "DELETE",
-        headers: { Authorizaton: `Bearer ${bearer}` },
-      });
-    })
+        headers: { Authorization: `Bot ${applicationSecret}` },
```
Comment (alvgaona) [inline src/setup.ts:23]:
```suggestion
        headers: { Authorization: `Bot ${token}` },
```

File: src/setup.ts [hunk 11]
```diff
@@ -1,93 +1,53 @@
...
-    commands: [ApplicationCommand, InteractionHandler][];
-  },
-  bearer: any
-): Promise<Response> => {
+type createCommandsArgs = {
+  applicationId: string;
+  guildId?: string;
+  commands: Command<any>[];
+};
+const createCommands = async ({ applicationId, guildId, commands }: createCommandsArgs, applicationSecret: string): Promise<Response> => {
```
Comment (alvgaona) [inline src/setup.ts:34]:
```suggestion
const createCommands = async ({ applicationId, guildId, commands }: createCommandsArgs, token: string): Promise<Response> => {
```

File: src/setup.ts [hunk 12]
```diff
@@ -1,93 +1,53 @@
...
-      body: JSON.stringify(command),
-      headers: { "Content-Type": "application/json", Authorization: `Bearer ${bearer}` },
-    });
-
     try {
-      const response = await fetch(request);
+      const response = await fetch(url, {
+        method: "POST",
+        body: JSON.stringify(command),
+        headers: { "Content-Type": "application/json", Authorization: `Bot ${applicationSecret}` },
```
Comment (alvgaona) [inline src/setup.ts:42]:
```suggestion
        headers: { "Content-Type": "application/json", Authorization: `Bot ${token}` },
```

File: src/handler.ts [hunk 13]
```diff
@@ -1,32 +1,44 @@
...
 const router = Router();
 
+export type Command<DataType extends InteractionDataType | void = void> = [
+  ApplicationCommand,
+  DataType extends InteractionDataType ? CommandInteractionHandlerWithData<DataType> : CommandInteractionHandler
+];
+
 export type Application = {
   applicationId: string;
   applicationSecret: string;
```
Comment (alvgaona) [inline src/handler.ts:19]:
```suggestion
  botToken: string;
```

**Review by alvgaona — CHANGES_REQUESTED** [review]
@Inrixia finally the review. Sorry for the long wait, we're definitely improving this! Thanks a lot.

**Comment by Inrixia** [comment]
Had a quick look at your comments. All looks good will sort this tomorrow

**Comment by Inrixia** [comment]
Surprise, more commits.

I went ahead and did some tests and realized that this library was using Node.js libraries *naughty* for encryption so I have replaced them allowing for running code directly on workers without the need for painful transpiling.

I have also exported all the `discord-api-types` directly from the package so people won't need to go looking for them or look to install the package themselves for them.

Finally I added a proper example both in the updated readme and in the new examples folder.

Please take a look and lemme know if you need any other changes before merging

File: examples/hello/.gitignore [hunk 14]
```diff
@@ -0,0 +1,2 @@
+node_modules
+wrangler.toml
```
Comment (alvgaona) [inline examples/hello/.gitignore:2]: The file needs an empty line at the EOF.

File: examples/hello/package.json [hunk 15]
```diff
@@ -0,0 +1,12 @@
+{
+  "main": "src/index.ts",
+  "author": "@Inrixia",
+  "dependencies": {
+    "cloudflare-discord-bot": "../../",
```
Comment (alvgaona) [inline examples/hello/package.json:5]: We'll change the name of the library to something nicer (follow-up PR).

File: examples/hello/src/hug.ts [hunk 16]
```diff
@@ -0,0 +1,22 @@
+import { Command, InteractionDataType, InteractionResponseType } from "cloudflare-discord-bot";
+
+export const hug: Command<InteractionDataType.ChatInput> = [
+  {
+    name: "hello",
+    description: "Ask your bot for a friendly hug!",
+  },
+  async (interaction) => {
+    const userId = interaction?.member?.user.id;
+    if (userId === undefined) throw new Error(`User is undefined`);
```
Comment (alvgaona) [inline examples/hello/src/hug.ts:10]: I think we can do `if (!userId)` to be a little broader. Like `""`, `null`, `undefined` would work as a falsy value.

File: src/setup.ts [hunk 17]
```diff
@@ -1,93 +1,52 @@
...
 const resolveCommandsEndpoint = (applicationId: string, guildId?: string): string => {
-  const url = `https://discord.com/api/v8/applications/${applicationId}`;
-
-  if (guildId) {
-    return `${url}/guilds/${guildId}/commands`;
-  }
-
-  return `${url}/commands`;
+  if (guildId !== undefined) return RouteBases.api + Routes.applicationGuildCommands(applicationId, guildId);
+  return RouteBases.api + Routes.applicationCommands(applicationId);
```
Comment (alvgaona) [inline src/setup.ts:7]:
```suggestion
  if (guildId) return RouteBases.api + Routes.applicationGuildCommands(applicationId, guildId);
  return RouteBases.api + Routes.applicationCommands(applicationId);
```

Comment (alvgaona) [reply src/setup.ts:7]: Same idea as above.

File: src/interaction.ts [hunk 18]
```diff
@@ -1,61 +1,91 @@
...
 
-    if (!isValid) throw new Error("Invalid request");
-  };
+import { InteractionType } from "./types";
+
+export enum InteractionDataType {
+  // APIChatInputApplicationCommandInteractionData
+  ChatInput,
+  // APIContextMenuInteractionData
+  ContextMenu,
```
Comment (alvgaona) [inline src/interaction.ts:21]: I think we need to replace these names as they don't exist anymore in the file.

File: src/interaction.ts [hunk 19]
```diff
@@ -1,61 +1,91 @@
...
+export type CommandInteractionHandler = (interaction: APIApplicationCommandInteraction, ...extra: any) => InteractionResponse;
+export type ComponentInteractionHandler = (interaction: Partial<APIMessageComponentInteraction>, ...extra: any) => InteractionResponse;
+
+const validateRequest = async (request: Request, publicKey: Uint8Array) => {
+  const signature = request.headers.get("x-signature-ed25519");
+  const timestamp = request.headers.get("x-signature-timestamp");
+
+  if (signature === null || timestamp === null) return false;
+
+  const encoder = new TextEncoder();
```
Comment (alvgaona) [inline src/interaction.ts:44]: Neat!

File: src/handler.ts [hunk 20]
```diff
@@ -1,32 +1,50 @@
...
+import type { CommandInteractionHandler, ComponentInteractionHandler } from "./interaction";
+import type { PartialAPIApplicationCommand } from "./types";
 
 const router = Router();
 
+export type Command<DataType extends InteractionDataType | void = void> = [
+  PartialAPIApplicationCommand,
+  DataType extends InteractionDataType ? CommandInteractionHandlerWithData<DataType> : CommandInteractionHandler
+];
+
```
Comment (alvgaona) [inline src/handler.ts:16]: What do you think of creating a `Component` type similar to this?

```typescript
type Component = [
 string,
 ComponentInteractionHandler
]
```

File: src/handler.ts [hunk 21]
```diff
@@ -1,32 +1,50 @@
...
 export type Application = {
   applicationId: string;
-  applicationSecret: string;
+  botToken: string;
   publicKey: string;
   guildId?: string;
-  commands: [ApplicationCommand, InteractionHandler][];
+  commands: Command<any>[];
```
Comment (alvgaona) [inline src/handler.ts:22]: Should we change `any` to a more restrictive type? WDYT?

**Review by alvgaona — CHANGES_REQUESTED** [review]
Nice changes @Inrixia! Left a few comments, nothing to big and a few questions for you.

Regarding the example: can we move away the example to a different repository? If you'd like to have the example linked to this repo we could use `git submodules`. 

I have created this https://github.com/discordcf/hello-bot to make it public once we make a new release of the framework.

**Comment by Inrixia** [comment]
@alvgaona great comments. Will update when I get home

File: src/handler.ts (hunk 21 above)
Comment (Inrixia) [reply src/handler.ts:22]: I always avoid using `any` where possible but in this case narrowing is not possible as different command signatures are incompatible with each other.

File: src/handler.ts (hunk 20 above)
Comment (Inrixia) [reply src/handler.ts:16]: There is no point as `ComponentInteractionHandler` is already exported to the user. Also the type you specified here is wrong, Components use a dict not an array.

File: src/interaction.ts (hunk 18 above)
Comment (Inrixia) [reply src/interaction.ts:21]: Both names are still used?
`ChatInput` is for API**ChatInput**ApplicationCommandInteractionData
`ContextMenu` is for API**ContextMenu**InteractionData

File: examples/hello/src/hug.ts (hunk 16 above)
Comment (Inrixia) [reply examples/hello/src/hug.ts:10]: This is a bad idea, you always want strict checks. The typings say that you will only get undefined. Not doing strict checks can leave to unexpected behaviour.

File: src/setup.ts (hunk 17 above)
Comment (Inrixia) [reply src/setup.ts:7]: See above

**Comment by Inrixia** [comment]
> Regarding the example: can we move away the example to a different repository? If you'd like to have the example linked to this repo we could use `git submodules`. Additionally, we could automate the update every time a push event is emitted from the main branch in that repo (but this is an optimization).

This is over engineering, having multiple repos for examples etc is more difficult all round especially for users to find them (most of the time they won't) and makes maintaining them harder. Also the fact that you suggest adding automation to update the main repo when changes are pushed from the secondary is a pretty clear indicator that it belongs in the main repo.

Speaking from experience of converting a major project to submodules in a github org (at least 8 repos + 2 helper repos and 1 "main" repo) and back again submodules look nice but can be a nightmare in practise, they have their use but this is definitely not it and I personally would advocate against using them as monorepos are much much easier to deal with.

File: src/interaction.ts (hunk 18 above)
Comment (alvgaona) [reply src/interaction.ts:21]: Yes, my bad! Don't know what I was thinking.

**Review by alvgaona — CHANGES_REQUESTED** [review]
Just reopened a conversation. Let's add that, I'm afraid that for some reason Discord API changes for some reason and breaks us.

**Comment by alvgaona** [comment]
> > Regarding the example: can we move away the example to a different repository? If you'd like to have the example linked to this repo we could use `git submodules`. Additionally, we could automate the update every time a push event is emitted from the main branch in that repo (but this is an optimization).
> 
> This is over engineering, having multiple repos for examples etc is more difficult all round especially for users to find them (most of the time they won't) and makes maintaining them harder. Also the fact that you suggest adding automation to update the main repo when changes are pushed from the secondary is a pretty clear indicator that it belongs in the main repo.
> 
> Speaking from experience of converting a major project to submodules in a github org (at least 8 repos + 2 helper repos and 1 "main" repo) and back again submodules look nice but can be a nightmare in practise, they have their use but this is definitely not it and I personally would advocate against using them as monorepos are much much easier to deal with.

@Inrixia I'm not suggesting we use a repo for each example bot. At the moment this is the only example we've got but we could do a monorepo for the examples, I'm in favor of doing that. I would prefer we keep this core repo as lean as possible, and just link to the examples repo in the documentation. If we don't want to do submodules, I'm fine with it; although, I do insist we keep the examples in a repo named `discordcf/example-bots` or sth.

**Comment by Inrixia** [comment]
@alvgaona If we don't have the examples embedded in the repo then submodules is a good intermediary as at least on github it links nicely on the web ui.

**Comment by Inrixia** [comment]
Also you should rename all the example repos to be prefixed with `examples-` or some other tag so it's easier to find them, currently it's not obvious at all what they are.

**Review by alvgaona — APPROVED** [review]

**Pull request merged**